### PR TITLE
Intergrate LibMEI code with the Verovio repository

### PR DESCRIPTION
**Pull request by lpugin**
The parse schema is in:
`./libmei/tools/parseschema2.py`

The output directory is
`./libmei/dist`

Additional files are in
`./libmei/addons`

The script as additional option (see README.md) to specify another output directory. There is also an option to generate element classes and to change the CPP namespace (WIP).

### CODE REVIEW

File: libmei/tools/parseschema2.py [hunk 1]
```diff
@@ -22,23 +22,24 @@
 # WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 import sys
+from typing import Pattern
```
Comment (rettinghaus) [inline libmei/tools/parseschema2.py:25]: typing is available from Python 3.5.

File: libmei/tools/parseschema2.py [hunk 2]
```diff
@@ -22,23 +22,24 @@
...
 
 if sys.version_info < (3, 4):
```
Comment (rettinghaus) [inline libmei/tools/parseschema2.py:27]: Still supporting outdated versions?

File: libmei/tools/parseschema2.py [hunk 3]
```diff
@@ -76,57 +81,54 @@ def __init__(self, oddfile):
 
         # lg.debug(self.data_lists)
 
+    @elapsedtime
     def get_elements(self):
         """Retrieves all defined elements from the schema."""
```
Comment (rettinghaus) [inline libmei/tools/parseschema2.py:86]:
```suggestion
        """Retrieve all defined elements from the schema."""
```
PEP 257

File: libmei/tools/parseschema2.py [hunk 4]
```diff
@@ -279,31 +270,49 @@ def strpatt(self, string: str) -> str:
...
-            return re.sub("[\s\t]+", " ", desc.xpath("string()"))
-        else:
+        # NB: `find()` doesn't support normal xpath value substitution, so we use string formatting instead
+        desc = self.schema.find(f".//tei:attDef[@ident='{att_name}']/tei:desc", namespaces=TEI_NS)
+        if desc is None:
             return ""
+        return re.sub(WHITESPACE_REGEX, " ", desc.xpath("string()"))
 
     def get_elem_desc(self, elem_name: str) -> str:
         """Returns the documentation string for an element by name."""
```
Comment (rettinghaus) [inline libmei/tools/parseschema2.py:288]:
```suggestion
        """Return the documentation string for an element by name."""
```
PEP 257

File: libmei/tools/parseschema2.py [hunk 5]
```diff
@@ -279,31 +270,49 @@ def strpatt(self, string: str) -> str:
 
     def strpdot(self, string: str) -> str:
         """Returns a version of the string without any dots."""
-        return "".join([n for n in string.split(".")])
+        return "".join(string.split("."))
 
     def cc(self, att_name: str) -> str:
         """Returns a CamelCasedName version of attribute.case.names."""
```
Comment (rettinghaus) [inline libmei/tools/parseschema2.py:276]:
```suggestion
        """Return a CamelCasedName version of attribute.case.names."""
```
PEP 257

File: libmei/tools/parseschema2.py [hunk 6]
```diff
@@ -279,31 +270,49 @@ def strpatt(self, string: str) -> str:
...
         return "".join([n[0].upper() + n[1:] for n in att_name.split(".")])
 
     def get_att_desc(self, att_name: str) -> str:
         """Returns the documentation string for an attribute by name."""
```
Comment (rettinghaus) [inline libmei/tools/parseschema2.py:280]:
```suggestion
        """Return the documentation string for an attribute by name."""
```
PEP 257

File: libmei/tools/cpp.py [hunk 7]
```diff
@@ -520,170 +523,149 @@ class AttModule {{
...
-        lg.debug("Config file could not be found")
-        return
-
-    f = config.open("r")
-    CONFIG = yaml.load(f, Loader=yaml.FullLoader)
 
-
-def vrv_get_att_config(module, gp, att):
-    if not module in CONFIG["modules"] or not gp in CONFIG["modules"][module]:
+def vrv_get_att_config(module, gp, att) -> Optional[dict]:
```
Comment (rettinghaus) [inline libmei/tools/cpp.py:550]: missing type hints

File: libmei/tools/cpp.py [hunk 8]
```diff
@@ -520,170 +523,149 @@ class AttModule {{
...
-    return "{0}_{1}".format(vrv_member_cc(att.replace("att.", "")), vallist.upper().replace(".", "").replace(":", ""))
+def vrv_getformattedvallist(att: str, vallist: str) -> str:
+    pfx: str = vrv_member_cc(att.replace("att.", ""))
+    sfx: str = vallist.upper().replace(".", "").replace(":", "")
+    return f"{pfx}_{sfx}"
 
 
-def vrv_getatttype(schema, module, gp, aname):
+def vrv_getatttype(schema, module: str, gp: str, aname: str) -> str:
     """Returns the attribute type for element name, or string if not detectable."""
```
Comment (rettinghaus) [inline libmei/tools/cpp.py:596]:
```suggestion
    """Return the attribute type for element name, or string if not detectable."""
```
PEP 257

File: libmei/tools/cpp.py [hunk 9]
```diff
@@ -520,170 +523,149 @@ class AttModule {{
...
-    ref = definition[0].xpath("tei:datatype/tei:dataRef/@key|tei:datatype/rng:ref/@name", gp=gp, name=aname, namespaces=TEI_RNG_NS)
+    ref = definition[0].xpath("./tei:datatype/tei:dataRef/@key|tei:datatype/rng:ref/@name",
+                              namespaces=TEI_RNG_NS)
     if ref:
-        return vrv_getformattedtype("{0}".format(ref[0]))
+        return vrv_getformattedtype(f"{ref[0]}")
+
     # Finally from val lists
-    vl = definition[0].find("tei:valList[@type=\"closed\"]", namespaces=TEI_RNG_NS)
+    vl = definition[0].find('./tei:valList[@type="closed"]', namespaces=TEI_RNG_NS)
```
Comment (rettinghaus) [inline libmei/tools/cpp.py:628]: consistent use of quotes would be preferred

File: libmei/tools/cpp.py [hunk 10]
```diff
@@ -1076,166 +1060,161 @@ def __create_element_classes(cpp_ns, schema, outdir):
...
-        if "std::string" in element_output:
-            outvars["includes"] += "#include <string>\n"
 
-        fullout = CLASS_FILE_H.format(**outvars)
-        f_element_class_h = open(Path(outdir, f"{module.lower()}.h"), "w")
-        f_element_class_h.write(fullout)
-        f_element_class_h.close()
-        lg.debug(f"\tCreated {module.lower()}.h")
+        with open(Path(outdir, f"{module.lower()}.h"), "w") as f_element_class_h:
+            f_element_class_h.write(CLASS_FILE_H.format_map(outvars))
```
Comment (rettinghaus) [inline libmei/tools/cpp.py:1089]: No need to open the Path object here, use `Path.write_text`.

**Review by rettinghaus — CHANGES_REQUESTED** [review]

File: libmei/tools/parseschema2.py [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+lg.addHandler(h)
+
+# globals
+TEI_NS = {"tei": "http://www.tei-c.org/ns/1.0"}
+TEI_RNG_NS = {"tei": "http://www.tei-c.org/ns/1.0",
+              "rng": "http://relaxng.org/ns/structure/1.0"}
+NAMESPACES = {'xml': 'http://www.w3.org/XML/1998/namespace',
+              'xlink': 'http://www.w3.org/1999/xlink'}
+
+WHITESPACE_REGEX: Pattern = re.compile(r"[\s\t]+")
```
Comment (rettinghaus) [inline libmei/tools/parseschema2.py:50]: now typing.Pattern is missing

Comment (ahankinson) [reply libmei/tools/parseschema2.py:50]: I did that on purpose, just to see if you would notice.

Comment (ahankinson) [reply libmei/tools/parseschema2.py:50]: I think it's fair to say that this is still WIP (hence Draft) so not quite ready for review.

File: libmei/tools/cpp.py [hunk 12]
```diff
@@ -1029,17 +1029,25 @@ def create_element_classes(cpp_ns: str, schema, outdir: Path):
...
                 att_str = schema.cc(schema.strpatt(attribute))
                 element_att_classes.append(f", Att{att_str}()")
-                element_read.append(f"\n    hasAttribute = (Read{att_str}(element) || hasAttribute);")
+                element_read.append(f"\n    hasAttribute = (Read{att_str}(element, removeAttr) || hasAttribute);")
                 element_write.append(f"\n    hasAttribute = (Write{att_str}(element) || hasAttribute);")
                 element_reset.append(f"\n    Reset{att_str}();")
 
+            readParam = "pugi::xml_node, bool"
+            writeParam = "pugi::xml_node"
+            if len(element_read):
```
Comment (ahankinson) [inline libmei/tools/cpp.py:1038]: You can just do `if element_read:` since an empty list is Falsey

File: libmei/tools/cpp.py [hunk 13]
```diff
@@ -1029,17 +1029,25 @@ def create_element_classes(cpp_ns: str, schema, outdir: Path):
 
                 att_str = schema.cc(schema.strpatt(attribute))
                 element_att_classes.append(f", Att{att_str}()")
-                element_read.append(f"\n    hasAttribute = (Read{att_str}(element) || hasAttribute);")
+                element_read.append(f"\n    hasAttribute = (Read{att_str}(element, removeAttr) || hasAttribute);")
                 element_write.append(f"\n    hasAttribute = (Write{att_str}(element) || hasAttribute);")
                 element_reset.append(f"\n    Reset{att_str}();")
 
+            readParam = "pugi::xml_node, bool"
```
Comment (ahankinson) [inline libmei/tools/cpp.py:1036]: should probably be `read_param` and `write_param`

**Comment by craigsapp** [comment]
Regarding the rename of generatecode.py to cpp.py: Note that there is a command-line tool called `cpp` (C preprocessor) with which the new name may cause confusions with.

**Comment by ahankinson** [comment]
You don't actually ever call cpp.py -- it's just a library of CPP formatting output.

**Comment by lpugin** [comment]
Looks good to me now

**Comment by ahankinson** [comment]
@rettinghaus you may now tell me everything that is wrong. :-D

File: include/vrv/interface.h [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * Virtual method returning the InterfaceId of the interface.
+     * Needs to be overridden in child classes.
+     */
+    virtual InterfaceId IsInterface() const { return INTERFACE; }
+
+private:
+    /**
+     * A vector for storing all the MEI att classes grouped in the interface
```
Comment (rettinghaus) [inline include/vrv/interface.h:68]: End with full stop.

File: include/vrv/interface.h [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+     * @name Constructors, destructors, and other standard methods
+     * Reset method reset all attribute classes
+     */
+    ///@{
+    Interface(){};
+    virtual ~Interface(){};
+    ///@}
+
+    /**
+     * Method for registering an MEI att classes in the interface.
```
Comment (rettinghaus) [inline include/vrv/interface.h:45]: Something not quite right here

File: include/vrv/interface.h [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    void RegisterInterfaceAttClass(AttClassId attClassId) { m_interfaceAttClasses.push_back(attClassId); }
+
+    /**
+     * Method for obtaining a pointer to the attribute class vector of the interface
```
Comment (rettinghaus) [inline include/vrv/interface.h:50]: End with full stop

File: libmei/tools/cpp.py [hunk 17]
```diff
@@ -0,0 +1,1267 @@
...
+        return '""', ["StrToStr", "StrToStr"]
+    else:
+        if default is None:
+            default = vrv_get_type_default(attype)
+
+        cname = vrv_converter_cc(attype)
+        return default, [f"StrTo{cname}", f"{cname}ToStr"]
+
+
+def create_doctstr(text: str, indent: int = 0) -> str:
```
Comment (rettinghaus) [inline libmei/tools/cpp.py:668]: Why is this called docTstr?

File: libmei/tools/cpp.py [hunk 18]
```diff
@@ -0,0 +1,1267 @@
...
+    # Header
+    ###########################################################################
+    for module, elements in sorted(schema.element_structure.items()):
+        if not elements:
+            continue
+
+        element_output: list = []
+        includes: list = []
+
+        for element, atgroups in sorted(elements.items()):
```
Comment (rettinghaus) [inline libmei/tools/cpp.py:966]: Why not attgroups or (even better) att_groups?

File: libmei/tools/datatypes.yml [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+        data_STAFFREL()
+
+modules:
+    externalsymbols:
+        att.extsym:
+            glyphnum:
+                type: wchar_t
+                default: 0
+
+    gestural:
```
Comment (rettinghaus) [inline libmei/tools/datatypes.yml:196]: Do we keep "gestural" here instead of "ges"?

File: libmei/tools/schema.py [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+from re import Pattern
+
+from lxml import etree
+
+logging.basicConfig(format="[%(asctime)s] [%(levelname)8s] %(message)s (%(filename)s:%(lineno)s)")
+log = logging.getLogger('schemaparser')
+
+
+# globals
+TEI_NS = {"tei": "http://www.tei-c.org/ns/1.0"}
```
Comment (rettinghaus) [inline libmei/tools/schema.py:12]: Why keeping this?

File: libmei/tools/schema.py [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+            self.get_elements()
+
+        self.get_attribute_groups()
+        self.get_data_types_and_lists()
+        self.invert_attribute_group_structure()
+        self.set_active_modules()
+        # lg.debug(self.data_lists)
+
+    def get_elements(self):
+        """Retrieves all defined elements from the schema."""
```
Comment (rettinghaus) [inline libmei/tools/schema.py:47]: "Retrieve all"

File: libmei/tools/schema.py [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+
+            self.attribute_group_structure[group_module][group_name] = []
+            for attdef in attdefs:
+                if attdef.get("ident") == "id":
+                    continue
+                attname = self.__process_att(attdef)
+                self.attribute_group_structure[group_module][group_name].append(attname)
+
+    def get_data_types_and_lists(self):
+        compound_alternate = self.schema.xpath(".//tei:macroSpec[@type='dt' and .//tei:alternate[@minOccurs='1' and @maxOccurs='1']]",
```
Comment (rettinghaus) [inline libmei/tools/schema.py:109]: Docstring would be nice here.

File: libmei/tools/schema.py [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+        self.invert_attribute_group_structure()
+        self.set_active_modules()
+        # lg.debug(self.data_lists)
+
+    def get_elements(self):
+        """Retrieves all defined elements from the schema."""
+        elements: list[etree.Element] = self.schema.xpath("//tei:elementSpec", namespaces=TEI_NS)
+
+        for element in elements:
+            modname = element.get("module").split(".")[-1]
```
Comment (rettinghaus) [inline libmei/tools/schema.py:51]: Further down you use "*_name"

File: libmei/tools/schema.py [hunk 24]
```diff
@@ -0,0 +1,235 @@
...
+            # need a way to keep self-defined attributes:
+            selfattributes = []
+            attdefs: list[etree.Element] = element.xpath("./tei:attList/tei:attDef", namespaces=TEI_NS)
+            if not attdefs:
+                continue
+
+            for attdef in attdefs:
+                if attdef.get("ident") == "id":
+                    continue
+                attname = self.__process_att(attdef)
```
Comment (rettinghaus) [inline libmei/tools/schema.py:78]: "att_name"?

File: libmei/tools/schema.py [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+
+        if member_groupel.xpath("./tei:attList/tei:attDef", namespaces=TEI_NS):
+            resarr.append(member_groupel.get("ident"))
+
+        m2s = member_groupel.xpath("./tei:classes/tei:memberOf", namespaces=TEI_NS)
+        for mship in m2s:
+            self.__get_membership(mship, resarr)
+
+    def strpatt(self, string: str) -> str:
+        """Returns a version of the string with any leading att. stripped."""
```
Comment (rettinghaus) [inline libmei/tools/schema.py:210]: Here ans following "Return".

File: libmei/tools/schema.py [hunk 26]
```diff
@@ -0,0 +1,235 @@
...
+        return string.replace("att.", "")
+
+    def strpdot(self, string: str) -> str:
+        """Returns a version of the string without any dots."""
+        return "".join(string.split("."))
+
+    def cc(self, att_name: str) -> str:
```
Comment (rettinghaus) [inline libmei/tools/schema.py:217]: I think "cc_name" would be more speaking

**Review by rettinghaus — CHANGES_REQUESTED** [review]
Looks good overall!

File: libmei/tools/schema.py (hunk 26 above)
Comment (ahankinson) [reply libmei/tools/schema.py:232]: Possibly, but I think with the docstring and the pattern of naming other things `_cc` in the cpp.py file, I'll leave it like this for now.

File: libmei/tools/cpp.py (hunk 17 above)
Comment (ahankinson) [reply libmei/tools/cpp.py]: Fixed

File: libmei/tools/schema.py (hunk 23 above)
Comment (ahankinson) [reply libmei/tools/schema.py]: Fixed.

File: libmei/tools/schema.py (hunk 24 above)
Comment (ahankinson) [reply libmei/tools/schema.py]: Fixed

File: libmei/tools/schema.py (hunk 25 above)
Comment (ahankinson) [reply libmei/tools/schema.py]: Fixed (I think)

File: libmei/tools/schema.py (hunk 22 above)
Comment (ahankinson) [reply libmei/tools/schema.py:116]: Fixed.

**Pull request merged**